pixivpy3/api.py

```diff
@@ -144,28 +183,43 @@ def auth(self, username=None, password=None, refresh_token=None, headers=None):
             self.access_token = token.response.access_token
             self.refresh_token = token.response.refresh_token
         except json.JSONDecodeError:
-            raise PixivError('Get access_token error! Response: %s' % token, header=r.headers, body=r.text)
+            raise PixivError(
+                "Get access_token error! Response: %s" % token,
+                header=r.headers,
+                body=r.text,
+            )
 
         # return auth/token response
         return token
 
-    def download(self, url, prefix='', path=os.path.curdir, name=None, replace=False, fname=None,
-                 referer='https://app-api.pixiv.net/'):
+    def download(
+        self,
+        url,
+        prefix="",
+        path=os.path.curdir,
+        name=None,
+        replace=False,
+        fname=None,
+        referer="https://app-api.pixiv.net/",
+    ):
+        # type: (str, str, str, Optional[str], bool, Optional[Union[str, IO[bytes]]], str) -> bool
         """Download image to file (use 6.0 app-api)"""
-        if hasattr(fname, 'write'):
+        if hasattr(fname, "write"):
             # A file-like object has been provided.
             file = fname
         else:
             # Determine file path by parameters.
-            name = prefix + (name or fname or os.path.basename(url))
+            name = prefix + str(name or fname or os.path.basename(url))
             file = os.path.join(path, name)
             if os.path.exists(file) and not replace:
                 return False
 
-        with self.requests_call('GET', url, headers={'Referer': referer}, stream=True) as response:
+        with self.requests_call(
+            "GET", url, headers={"Referer": referer}, stream=True
+        ) as response:
             if isinstance(file, str):
-                with open(file, 'wb') as out_file:
+                with open(file, "wb") as out_file:
                     shutil.copyfileobj(response.raw, out_file)
-            else:
+            elif isinstance(file, io.IOBase):
```

Comment:
Similar as above, `file` doesn't need to be `IOBase`. And we should have an `else` branch to capture all cases.

Reply:
`isinstance(file, io.IOBase)` is needed to judge if `file` is `IO[bytes]` or `None`. Without this cond:

```shellsession
$ git diff
diff --git a/pixivpy3/api.py b/pixivpy3/api.py
index 3cc94f5..d762791 100644
--- a/pixivpy3/api.py
+++ b/pixivpy3/api.py
@@ -220,6 +220,7 @@ class BasePixivAPI(object):
             if isinstance(file, str):
                 with open(file, "wb") as out_file:
                     shutil.copyfileobj(response.raw, out_file)
-            elif isinstance(file, io.IOBase):
+            # elif isinstance(file, io.IOBase):
+            elif hasattr(file, "write"):
                 shutil.copyfileobj(response.raw, file)
         return True
$ mypy -p pixivpy3
pixivpy3/api.py:225: error: Argument 2 to "copyfileobj" has incompatible type "Optional[IO[bytes]]"; expected
"SupportsWrite[Any]"  [arg-type]
                    shutil.copyfileobj(response.raw, file)
                                                     ^
Found 1 error in 1 file (checked 5 source files)
```

```shellsession
$ git diff
diff --git a/pixivpy3/api.py b/pixivpy3/api.py
index 3cc94f5..f1bef2e 100644
--- a/pixivpy3/api.py
+++ b/pixivpy3/api.py
@@ -220,6 +220,8 @@ class BasePixivAPI(object):
             if isinstance(file, str):
                 with open(file, "wb") as out_file:
                     shutil.copyfileobj(response.raw, out_file)
-            elif isinstance(file, io.IOBase):
+            # elif isinstance(file, io.IOBase):
+            # elif hasattr(file, "write"):
+            else:
                 shutil.copyfileobj(response.raw, file)
         return True
$ mypy -p pixivpy3
pixivpy3/api.py:226: error: Argument 2 to "copyfileobj" has incompatible type "Optional[IO[bytes]]"; expected
"SupportsWrite[Any]"  [arg-type]
                    shutil.copyfileobj(response.raw, file)
                                                     ^
Found 1 error in 1 file (checked 5 source files)

Reply:
Can we suppress mypy check for this line? Just modifying the logic for the sake of type annotation doesn't quite make sense to me.

Reply:
6002e01